### DISCUSSION TITLE
How to link doctrine resource with API resource not mapped with doctrine

### DISCUSSION THREAD

**Discussion by KerianMM** [api-platform/core#5557] [Q&A] [answered]
Hello

[gist with 4 example files](https://gist.github.com/KerianMM/384384a1c9cad38d21960b096ef6cad5)

Our `ContractInternal` doctrine entity can be updated by API with a PUT on `RequestChange` resource.
A security step check if the change request is done on a contract owned by the current user.
If all is ok, the processor will apply changes on entity.

How to link doctrine resource with API resource not mapped with doctrine ?
Doctrine item provider cannot provide a non entity object

We understood that the operation's `stateOptions` is a free way to extend basic comportment, so we created the `MappedResourceOptions`.

But we still have some doubts.
May be we made a mistake about how should be use `stateOptions` ?

Without create provider for each, how to link doctrine resource with API resource not mapped with doctrine ?

Thank you for

**Comment by soyuka** [comment, marked as answer]
Hi,

Indeed, `stateOptions` or `extraProperties` are ways to extend our metadata to your liking.
We do this a lot when creating APIs, as it's very useful (for example to avoid serialization groups). For this to be automatic, we either do the mapping by hand or use a mapper (proposal for a new Symfony component: https://github.com/symfony/symfony/pull/51741), check also the [Symfony casts](https://symfonycasts.com/screencast/api-platform-extending/provider-return-dto) on the subject (also uses a [micro-mapper](https://github.com/SymfonyCasts/micro-mapper)). 

You can find an example with the proposed component at: https://github.com/soyuka/api-platform-doctrine-mapper/blob/main/src/State/BookProvider.php#L14. At some point I'd like this to be in the `core`. 

Now the last thing you need is to use your Provider on every resource or entity/resource tuple. To do so, use a Metadata Factory, like the one that works on doctrine entities inside the API Platform code: 

https://github.com/api-platform/core/blob/e895b69a04cb52eb6eb030bcc51a1e143dd357cb/src/Doctrine/Orm/Metadata/Resource/DoctrineOrmResourceCollectionMetadataFactory.php#L88-L105
